### PR TITLE
115 long running queries

### LINKED ISSUE #115: Long running queries/expired recalls
This has been somewhat of an ongoing problem on the Diamond ICAT system that has been investigated previously but never resolved.

The problem occurs when a user submits a download/restore request via TopCAT. TopCAT then passes the request on to the IDS in the form of a prepareData call and waits for the IDS to respond with a preparedID. What happens next is that the IDS takes longer than the 30 minute timeout to respond, so TopCAT does not get a preparedID and cannot progress the request. (Just as a side note increasing the 30 minute timeout is not really a suitable/sustainable option because some of the requests take hours to complete and the request really should not take more than 30 minutes).

The problem tends to occur for InstrumentScientists who have access to a lot of data but has also been seen for InvestigationUsers who also have access to many Investigations.

From a recent investigation, our database admin identified the following query which had been running on the database for a long time:

`SELECT *
FROM
(SELECT a.*, ROWNUM rnum
FROM
(SELECT t0.ID AS a1, t0.NAME AS a2, t0.LOCATION AS a3, t0.CREATE_ID AS a4, t0.MOD_ID AS a5
FROM DATAFILE t0
WHERE ((((t0.DATASET_ID = :1 ) AND (t0.LOCATION IS NOT NULL)) AND (t0.ID BETWEEN :2 AND :3 )) AND ((t0.ID IN
(SELECT t1.ID
FROM INVESTIGATION t3, DATASET t2, DATAFILE t1
WHERE ((t3.VISIT_ID IN (:4 , :5 )) AND ((t2.ID = t1.DATASET_ID) AND (t3.ID = t2.INVESTIGATION_ID)))) OR t0.ID IN
(SELECT DISTINCT t4.ID
FROM INVESTIGATIONINSTRUMENT t7, INVESTIGATION t6, DATASET t5, DATAFILE t4, USER_ t10, INSTRUMENTSCIENTIST t9, INSTRUMENT t8
WHERE ((t10.NAME = :6 ) AND ((((((t5.ID = t4.DATASET_ID) AND (t6.ID = t5.INVESTIGATION_ID)) AND (t7.INVESTIGATION_ID = t6.ID)) AND (t8.ID = t7.INSTRUMENT_ID)) AND (t9.INSTRUMENT_ID = t8.ID)) AND (t10.ID = t9.USER_ID))))) OR t0.ID IN
(SELECT DISTINCT t11.ID
FROM INVESTIGATION t13, DATASET t12, DATAFILE t11, USER_ t15, INVESTIGATIONUSER t14
WHERE ((t15.NAME = :7 ) AND ((((t12.ID = t11.DATASET_ID) AND (t13.ID = t12.INVESTIGATION_ID)) AND (t14.INVESTIGATION_ID = t13.ID)) AND (t15.ID = t14.USER_ID))))))) a
WHERE ROWNUM <= :8 )
WHERE rnum > :9
`

This is the result of the ICAT query at the following line in the manyDfs method of DataSelection being executed:
https://github.com/icatproject/ids.server/blob/v1.11.0/src/main/java/org/icatproject/ids/DataSelection.java#L247

The query in the code is simple enough but by the time it has had the InstrumentScientist and InvestigationUser "where" clauses added (along with another rule we have for some open data) the final query is rather complicated. 

Along with the complexity of the query, the number of results returned can vary between 1 and 250,000 (the maxEntities value of the ICAT) so it is thought that Oracle struggles to select an appropriate execution plan for the query.

The workaround at the moment is for the service manager to spot in the TopCAT admin tab when there is a user download in this state and to use a specially prepared script which generates a preparedID from the IDS using the login details for an ICAT root user and inserts it into TopCAT. Using a root user account means that none of the ICAT rules get added to the ICAT query and the final query that goes to the database is almost as simple as the query in the IDS code. Thus the preparedID is generated extremely quickly.

### PR DESCRIPTION
**Pull request by kevinphippsstfc**
Closes #115 

The solution is based around the use of the reader account (already used by the Tidier etc. in the IDS) and a new property useReaderForPerformance which must be set for this to be used. In addition, there must be a PublicStep defined from Dataset to datafiles, which then allows use of the reader account for doing the expensive Datafiles queries once it has been confirmed that the user has access to the parent Dataset.

If the property is not set or the PublicStep does not exist then the behaviour will be as before, with the user's session used to lookup both the Datasets and the Datafiles.

### CODE REVIEW

File: src/test/resources/test.properties [hunk 1]
```diff
@@ -1,12 +1,12 @@
-# These credentials must have permission to read and download
-# the dataset and datafiles specified below.
-authorizedIcatUsername=root
-authorizedIcatPassword=password
+# login credentials for a root user
+login.root = db username root password password
```
Comment (RKrahl) [inline src/test/resources/test.properties:2]: Appreciate this tidying up!

File: src/test/java/org/icatproject/ids/IcatReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    private void setupPropertyHandler() {
+        MockitoAnnotations.initMocks(this);
+        when(mockedPropertyHandler.getIcatService()).thenReturn(icatService);
+        when(mockedPropertyHandler.getReader()).thenReturn(readerCreds);
+    }
+
+    @Test
+    public void testIcatReaderCreation() throws IcatException_Exception {
+        IcatReader icatReader = new IcatReader(mockedPropertyHandler);
+        assertNotNull("An ICAT session ID was expected", icatReader.getSessionId());
```
Comment (RKrahl) [inline src/test/java/org/icatproject/ids/IcatReaderTest.java:66]: This requires the user `reader` to be added in the authentication plugin. Until now, the documentation on how to run the tests refer to the corresponding documentation for `icat.server` which in turn instructs to add the users `notroot`, `piOne`, `piTwo`, `root`, `guest`, and `CIC`.

File: src/site/xhtml/installation.xhtml.vm [hunk 3]
```diff
@@ -164,6 +164,13 @@
 			directory. The file ids.logback.xml.example may be renamed to
 			ids.logback.xml to get started.</dd>
 
+		<dt>useReaderForPerformance</dt>
+		<dd>Optional. If true allows the IDS to make use of the reader account
+			to improve query performance. Use of this option will always be
+			subject to an additional check that the user would be otherwise
+			allowed to perform the query, for example, that a PublicStep exists
+			and access to the parent entity is allowed.</dd>
+
```
Comment (RKrahl) [inline src/site/xhtml/installation.xhtml.vm:173]: If we drop the check on `PublicStep` as suggested in the comment in `IcatReader.java` we would need to change that comment here that setting `useReaderForPerformance` will effectively imply a public step.

**Review by RKrahl — CHANGES_REQUESTED** [review]
Many thanks for submitting this! I tested it and it works fine. Overall, I'm happy with the changes.

I left some comments on minor issues.  Most importantly, I suggest dropping the check for the `PublicStep` which would also simplify things, but would also require amending the documentation.

File: pom.xml [hunk 4]
```diff
@@ -94,6 +94,13 @@
 			<scope>test</scope>
 		</dependency>
 
+		<dependency>
+			<groupId>org.mockito</groupId>
+			<artifactId>mockito-core</artifactId>
+			<version>2.28.2</version>
+			<scope>test</scope>
+		</dependency>
+
```
[the remark this answers is not in the document]
Comment (kevinphippsstfc) [reply pom.xml:103]: Thanks for spotting this. I've now upgraded to the current version and there are some minor code changes to go with that. Sorry but I don't know about using Mockito with JUnit 4 vs JUnit 5.

File: src/main/java/org/icatproject/ids/PropertyHandler.java [hunk 5]
```diff
@@ -244,6 +245,8 @@ private PropertyHandler() {
 				logger.info("'log.list' entry not present so no JMS call logging will be performed");
 			}
 
+			useReaderForPerformance = props.getBoolean("useReaderForPerformance", false);
+
```
[the remark this answers is not in the document]
Comment (kevinphippsstfc) [reply src/main/java/org/icatproject/ids/PropertyHandler.java:249]: OK, I've added that.

**Review by RKrahl — APPROVED** [review]

**Comment by RKrahl** [comment]
Looks good for me now.  @stuartpullinger: do you still want to have a look before merging?

**Pull request merged**